### PR TITLE
Plugin Doc and Example Refactors for #5596

### PR DESCRIPTION
**Pull request by Spacesai1or**
This PR updates the plugin author and user docs as well as the example plugin code, according to the suggestion made in #5596

### CODE REVIEW

**Review by avkos — APPROVED** [review]

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md [hunk 1]
```diff
@@ -7,9 +7,9 @@ sidebar_label: 'Plugin Authors'
 
 This guide intends to provide the necessary context for developing plugins for web3.js.
 
-## Before Getting Started
-
-It's highly recommended you as the plugin author understand the limitations of TypeScript's module augmentation as described in the [main plugin guide](/docs/guides/web3_plugin_guide/), so you can communicate to your users that they are responsible for augmenting the class interface they register your plugin with if they desire to have type support when using your plugin. Ideally this could be solved for by the plugin author, or better yet web3.js, but so far a better solution is unknown - if you have any ideas, please [create an issue](https://github.com/web3/web3.js/issues/new/choose) and help us improve web3.js' UX.
+:::caution
+To provide type safety and hinting for your plugin users, please refer to the [Setting Up Module Augmentation](/docs/guides/web3_plugin_guide/plugin_authors#setting-up-module-augmentation) section for how to augment the `Web3Context` module to enable typing features for your plugin.
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_authors.md]: Tiny suggestion:
```suggestion
To provide type safety and IntelliSense for your plugin users, please refer to the [Setting Up Module Augmentation](/docs/guides/web3_plugin_guide/plugin_authors#setting-up-module-augmentation) section for how to augment the `Web3Context` module to enable typing features for your plugin.
```

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md [hunk 2]
```diff
@@ -98,14 +93,9 @@ Below is representing a plugin user's code that does not configure an Ethereum p
...
 
-declare module 'web3-core' {
-	interface Web3Context {
-		customRpcMethods: CustomRpcMethodsPlugin;
-	}
-}
+import { CustomRpcMethodsPlugin } from './custom_rpc_methods_plugin';
 
 const web3Context = new Web3Context();
 web3Context.registerPlugin(new CustomRpcMethodsPlugin());
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_authors.md:111]: This code snippet and its section seem to need to be moved up before the title `Using the Inherited Web3Context`, doesn't it?

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md [hunk 3]
```diff
@@ -210,3 +200,177 @@ public link(parentContext: Web3Context) {
 	this._contract.link(parentContext);
 }
 ```
+
+## Setting Up Module Augmentation
+
+In order to provide typing support for your plugin when it's registered by the user, we must [augment](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) the `Web3Context` module. In simpler terms, we're making TypeScript aware that we are modifying the interface of `Web3Context`, and any class that extends it, to include the interface of your plugin (i.e. your plugin's added methods, properties, etc.). A good tutorial that further explains the topic can be found [here](https://www.digitalocean.com/community/tutorials/typescript-module-augmentation).
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_authors.md]: I think there is a mixed "we" and "you" here.
So either using "you" and "your":
```suggestion
In order to provide type safety and IntelliSense for your plugin when it's registered by the user, you must [augment](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) the `Web3Context` module. In simpler terms, you will be making TypeScript aware that you are modifying the interface of `Web3Context`, and any class that extends it, to include the interface of your plugin (i.e. your plugin's added methods, properties, etc.). A good tutorial that further explains the topic can be found [here](https://www.digitalocean.com/community/tutorials/typescript-module-augmentation).
```

Or using "we" and "our" if we refer to the plugin user as we are talking about ourselves (but it is not a good suggestion I think because it would confuse the reader):
```suggestion
In order to provide type safety and IntelliSense for our plugin when it's registered by the user, we must [augment](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) the `Web3Context` module. In simpler terms, we're making TypeScript aware that we are modifying the interface of `Web3Context`, and any class that extends it, to include the interface of our plugin (i.e. your plugin's added methods, properties, etc.). A good tutorial that further explains the topic can be found [here](https://www.digitalocean.com/community/tutorials/typescript-module-augmentation).
```

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md [hunk 4]
```diff
@@ -210,3 +200,177 @@ public link(parentContext: Web3Context) {
...
+
+### A Quick Disclaimer
+
+The `registerPlugin` method exists on the `Web3Context` class, so any class that `extends Web3Context` has the ability to add on your plugin's additional functionality to it's interface. By augmenting `Web3Context` to include your plugin's interface, we're essentially providing a blanket augmentation that adds you plugin's interface to **all** Web3 modules that extend `Web3Context` (i.e. `web3`, `web3-eth`, `web3-eth-contract`, etc.).
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_authors.md]: Here also there are "... your plugin's ..." and "...we're...".

```suggestion
The `registerPlugin` method exists on the `Web3Context` class, so any class that `extends Web3Context` has the ability to add your plugin's additional functionality to its interface. And, by augmenting `Web3Context` to include your plugin's interface, you're essentially providing a blanket augmentation that adds your plugin's interface to **all** Web3 modules that extend `Web3Context` (i.e. `web3`, `web3-eth`, `web3-eth-contract`, etc.).
```

And actually besides using "you" is more clear, if "we" is used it may confuse the reader that this would be done, or already done, by web3.js team.

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md [hunk 5]
```diff
@@ -210,3 +200,177 @@ public link(parentContext: Web3Context) {
...
+
+```typescript
+import { Web3Context } from 'web3-core';
+
+export { Web3Context };
+```
+
+### Re-declaring the Module
+
+Now we're going to tell TypeScript that we're interested in re-defining a module's (in this case `reexported_web3_context`) interface. In simpler terms, TypeScript is already aware of what methods and classes exist for each web3.js module, but when registering a plugin, we're adding additional methods and/or classes to the module's interface and TypeScript needs a little help understanding what's going to be available within the module after the plugin is registered.
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_authors.md]:
```suggestion
Now you're going to tell TypeScript that you're interested in re-defining a module's (in this case `reexported_web3_context`) interface. In simpler terms, TypeScript is already aware of what methods and classes exist for each web3.js module, but when registering a plugin, you're adding additional methods and/or classes to the module's interface and TypeScript needs a little help understanding what's going to be available within the module after the plugin is registered.
```

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md [hunk 6]
```diff
@@ -210,3 +200,177 @@ public link(parentContext: Web3Context) {
...
+// Now customRpcMethods (i.e. the pluginNamespace) is available
+// on our instance of Web3
+web3.customRpcMethods;
+```
+
+:::
+
+### Exporting Our Augmented Web3Context
+
+Lastly we just need to export our augmented `Web3Context` by adding the following after our module re-declaration:
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_authors.md]: To keep with the context.
```suggestion
Lastly, you just need to export our augmented `Web3Context` by adding the following after our module re-declaration:
```

File: docs/docs/guides/web3_plugin_guide/plugin_users.md [hunk 7]
```diff
@@ -29,9 +25,7 @@ Unless otherwise mentioned by the plugin author, installing a plugin should be a
 
 The `.registerPlugin` method is what we're going to be using to add a plugin to an instance of a class sourced from web3.js' modules (i.e. `Web3` or `Web3Eth`). This method only exists on classes that extend `Web3Context`, so it may not be available on every class you import from a Web3.js package.
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_users.md]:
```suggestion
The `.registerPlugin` method is what you're going to be using to add a plugin to an instance of a class sourced from web3.js' modules (i.e. `Web3` or `Web3Eth`). This method only exists on classes that extend `Web3Context`, so it may not be available on every class you import from a Web3.js package.
```

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md [hunk 8]
```diff
@@ -210,3 +200,177 @@ public link(parentContext: Web3Context) {
...
+	}
+}
+
+// Here is where we are beginning to augment the Web3Context module
+declare module './reexported_web3_context' {...}
+```
+
+### Adding our Plugin's Interface
+
+Now that TypeScript's aware that the interface of the `reexport_web3_context` module is going to be augmented, we add our changes. In this case, we're adding the interface of `SimplePlugin` to the interface of `Web3Context` which is what we're going to be calling `.registerPlugin` on:
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_authors.md]: Actually, "TypeScript`s" is not the same as "TypeScript is".  We can not abbreviate "is" with names. With names, the " 's " is used to show possession and ownership.

Also using you to keep with the context.

```suggestion
Now that TypeScript is aware that the interface of the `reexport_web3_context` module is going to be augmented, you can add our changes. In this case, you're adding the interface of `SimplePlugin` to the interface of `Web3Context` which is what the **plugin-user** is going to be calling `.registerPlugin` on:
```

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md [hunk 9]
```diff
@@ -210,3 +200,177 @@ public link(parentContext: Web3Context) {
...
+	}
+}
+
+declare module './custom_rpc_methods_plugin.ts' {
+	interface Web3Context {
+		simplePlugin: SimplePlugin;
+	}
+}
+
+// Here is where we are exporting our augmented Web3Context
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_authors.md]: To keep with the context:
```suggestion
// Here is where you are exporting your augmented Web3Context
```

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md [hunk 10]
```diff
@@ -210,3 +200,177 @@ public link(parentContext: Web3Context) {
...
+
+export class CustomRpcMethodsPlugin extends Web3PluginBase {
+	public pluginNamespace = 'customRpcMethods';
+
+	public someMethod() {
+		return 'someValue';
+	}
+}
+
+// Here is where we are beginning to augment the Web3Context module
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_authors.md]:
```suggestion
// Here is the augmentation of the Web3Context module
```
(This would also remove "we" to keep with the context)

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md [hunk 11]
```diff
@@ -210,3 +200,177 @@ public link(parentContext: Web3Context) {
...
+Now we're going to tell TypeScript that we're interested in re-defining a module's (in this case `reexported_web3_context`) interface. In simpler terms, TypeScript is already aware of what methods and classes exist for each web3.js module, but when registering a plugin, we're adding additional methods and/or classes to the module's interface and TypeScript needs a little help understanding what's going to be available within the module after the plugin is registered.
+
+We start with the following:
+
+```typescript
+// custom_rpc_methods_plugin.ts
+import { Web3PluginBase } from 'web3-core';
+
+// Here we are importing our re-exported Web3Context
+// from the above section
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_authors.md]:
```suggestion
// Here is the import of the re-exported `Web3Context` 
// that has been mentioned in the above section.
```
(This would also remove "we" to keep with the context)

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md [hunk 12]
```diff
@@ -210,3 +200,177 @@ public link(parentContext: Web3Context) {
...
+![web3 context augmentation side effect](./assets/web3_context_augmentation_side_effect.png 'web3Context augmentation side effect')
+
+The above screenshot shows intellisense thinking `.customRpcMethods.someMethod` is available to call on the instance of `Web3`, even though we haven't registered `CustomRpcMethodsPlugin` - running this code would result in an error.
+:::
+
+### Re-exporting Web3Context
+
+Currently TypeScript's module augmentation only supports named exports, so the first step in augmenting `Web3Context` is to re-export it as a named export. To do this we're going to create a `reexported_web3_context.ts` file (the name of this file can be whatever, but for the sake of this guide, we're going to assume it's named `reexported_web3_context.ts` and is located within the same directory as our `custom_rpc_methods_plugin.ts` file). The file contents should be as follows:
+
+```typescript
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_authors.md:263]:
```suggestion
```typescript
// reexported_web3_context.ts
```

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md [hunk 13]
```diff
@@ -210,3 +200,177 @@ public link(parentContext: Web3Context) {
...
+import { Web3Context } from 'web3-core';
+
+export { Web3Context };
+```
+
+### Re-declaring the Module
+
+Now we're going to tell TypeScript that we're interested in re-defining a module's (in this case `reexported_web3_context`) interface. In simpler terms, TypeScript is already aware of what methods and classes exist for each web3.js module, but when registering a plugin, we're adding additional methods and/or classes to the module's interface and TypeScript needs a little help understanding what's going to be available within the module after the plugin is registered.
+
+We start with the following:
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_authors.md]: I think it is not the "start" here. I think it is the main logic or the main file that contains the plugin class.

```suggestion
Here is the content of the source file that contains the main plugin class:
```
(This would also remove "we" to keep with the context)

**Review by Muhammad-Altabba — CHANGES_REQUESTED** [review]
Looks great to me except few questions and suggestions.

And regarding the file named `web3_context_augmentation_side_effect.png`. What I see in the screenshot is the result of the augmentation **main** effect, which we are seeking. And using the words `side_effect` would probably make some confusion. As we call a "side effect", usually, on something that is not intended to happen, but happened due to some side effect. So, I suggest to simply call the file `web3_context_augmentation.png`.

And it will also be nice if you screen through the unchanged files to ensure replacing "we" with "you" and "our" with "yours", in order to unify the writing style. As currently there is a mixed usage of them sometimes even within the same paragraph or section.

The last thing, other than the comments below, is that I suggest using the term `Plugin Writer` instead of `Plugin Author`. As this would be more meaningful from a linguistic point of view, up to my understanding of the two terms. And yes, this is not related to this MR. But if you find it good to be done, I think you could edit those few places as part of this MR.

Thanks,

**Comment by Spacesai1or** [comment]
> And regarding the file named `web3_context_augmentation_side_effect.png`. What I see in the screenshot is the result of the augmentation **main** effect, which we are seeking. And using the words `side_effect` would probably make some confusion. As we call a "side effect", usually, on something that is not intended to happen, but happened due to some side effect. So, I suggest to simply call the file `web3_context_augmentation.png`.

@Muhammad-Altabba The screenshot is in the context of IntelliSense still showing the plugin's functionality even if the plugin hasn't been registered to an instance of the Web3 class - This is not intended functionality, but is a repercussion of the plugin augmenting `Web3Context`

**Comment by Muhammad-Altabba** [comment]
> > And regarding the file named `web3_context_augmentation_side_effect.png`. What I see in the screenshot is the result of the augmentation **main** effect, which we are seeking. And using the words `side_effect` would probably make some confusion. As we call a "side effect", usually, on something that is not intended to happen, but happened due to some side effect. So, I suggest to simply call the file `web3_context_augmentation.png`.
> 
> @Muhammad-Altabba The screenshot is in the context of IntelliSense still showing the plugin's functionality even if the plugin hasn't been registered to an instance of the Web3 class - This is not intended functionality, but is a repercussion of the plugin augmenting `Web3Context`

Ok, I got your point now :smile: 
But I think about it this way: The screenshot shows the direct-effect, that is the main-effect, and that is almost-only-effect of the Module augmentation. And showing that the IntelliSense will happen, regardless of calling `registerPlugin`, is something not to be confused by what is the main-effect and what is the side-effect of the Module augmentation. 
So, actually having the IntelliSense work is the intended effect of the module augmentation. The intended effect means we do the Modul augmentation for only one reason and it works for that reason. We do the module augmentation to have IntelliSense and it works as intended. And having the IntelliSense work regardless of calling `registerPlugin` is something that does not alter what is the Module augmentation's intended purpose. 
So, I think you meant to call it `web3_context_augmentation_regardless_of_calling_registerPlugin.png` but that would be a long name though :smile: and `web3_context_augmentation.png` is good enough, I think.
I think that we can keep the file name as you named it. I just think that it would confuse the developers who read its name later. 
Or what do you think?

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md (hunk 2 above)
Comment (Spacesai1or) [reply docs/docs/guides/web3_plugin_guide/plugin_authors.md:111]: No it's meant to be where it is. It's adding additional context for the plugin author showing how their code could be affected by what the user does/n't do (e.g. this example shows a provider error being thrown from the plugin author's code because the plugin user didn't provide a provider). [This](https://github.com/web3/web3.js/pull/5621/commits/96c90445070f2751a1286597cd40fdc24b3d421e) commit does reword this section a little

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md (hunk 4 above)
Comment (Spacesai1or) [reply docs/docs/guides/web3_plugin_guide/plugin_authors.md]: Addressed by [this](https://github.com/web3/web3.js/pull/5621/commits/92402f68f7ab0d0102a495714fd6188ac5e04c2e) commit

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md (hunk 13 above)
Comment (spacesailor24) [reply docs/docs/guides/web3_plugin_guide/plugin_authors.md]: Addressed bu [this](https://github.com/web3/web3.js/pull/5621/commits/4cb635cb4a8489aaf3d0cd3b6479aad3764ed20e) commit

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md (hunk 11 above)
Comment (spacesailor24) [reply docs/docs/guides/web3_plugin_guide/plugin_authors.md]: Addressed bu [this](https://github.com/web3/web3.js/pull/5621/commits/4cb635cb4a8489aaf3d0cd3b6479aad3764ed20e) commit

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md (hunk 13 above)
Comment (Spacesai1or) [reply docs/docs/guides/web3_plugin_guide/plugin_authors.md]: Addressed by [this](https://github.com/web3/web3.js/pull/5621/commits/4cb635cb4a8489aaf3d0cd3b6479aad3764ed20e) commit

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md (hunk 11 above)
Comment (Spacesai1or) [reply docs/docs/guides/web3_plugin_guide/plugin_authors.md]: Addressed by [this](https://github.com/web3/web3.js/pull/5621/commits/4cb635cb4a8489aaf3d0cd3b6479aad3764ed20e) commit

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md (hunk 10 above)
Comment (Spacesai1or) [reply docs/docs/guides/web3_plugin_guide/plugin_authors.md]: Addressed by [this](https://github.com/web3/web3.js/pull/5621/commits/4cb635cb4a8489aaf3d0cd3b6479aad3764ed20e) commit

**Comment by Spacesai1or** [comment]
> And it will also be nice if you screen through the unchanged files to ensure replacing "we" with "you" and "our" with "yours", in order to unify the writing style. As currently there is a mixed usage of them sometimes even within the same paragraph or section.

@Muhammad-Altabba Addressed by [this](https://github.com/web3/web3.js/pull/5621/commits/73b57522e5bb093e9d1d4d4cd107ed185ac24e2e) commit

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md [hunk 14]
```diff
@@ -210,3 +201,177 @@ public link(parentContext: Web3Context) {
...
+}
+
+// Here is the declaration to TypeScript that you are
+// augmenting the imported module (i.e. ./reexported_web3_context)
+declare module './reexported_web3_context' {...}
+```
+
+### Adding Your Plugin's Interface
+
+Now that TypeScript is aware that the interface of the `reexport_web3_context` module is going to be augmented, you can add your plugin's interface. In this case, you're adding the interface of `SimplePlugin` to the interface of `Web3Context` which is what the **plugin-user** is going to be calling `.registerPlugin` on:
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_authors.md]: The name `SimplePlugin` here is not used elsewhere in this section. However, the one used in the same section below is  `CustomRpcMethodsPlugin`. 
Actually, there is a mixed usage of the name of the plugin in other sections as well. And I think this would confuse the reader. So could you unify the name across all sections and code snippets, except if there is a need for a specific reason?

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md [hunk 15]
```diff
@@ -210,3 +201,177 @@ public link(parentContext: Web3Context) {
...
+
+### A Quick Disclaimer
+
+The `registerPlugin` method exists on the `Web3Context` class, so any class that `extends Web3Context` has the ability to add your plugin's additional functionality to its interface. By augmenting `Web3Context` to include your plugin's interface, you're essentially providing a blanket augmentation that adds your plugin's interface to **all** Web3 modules that extend `Web3Context` (i.e. `web3`, `web3-eth`, `web3-eth-contract`, etc.).
+
+:::warning
+By augmenting `Web3Context` (and by extension all class interfaces that extend it), your plugin's interface will show up in things like intellisense for **all** Web3 modules that extend `Web3Context`, even if your plugin isn't registered - This is something worth making your users aware of, as they'll only be able to use your plugin if they register it with a Web3 class instance using `.registerPlugin`
+
+For context, here is a code example of this augmentation side-effect (the code in this example is further explained in the subsequent sections):
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_authors.md]: I would like to know what the others think about this also :smile:. Especially our other approver: @avkos 

```suggestion
For context, here is a code example of this augmentation that happens regardless of whether `registerPlugin` is called or not (the code in this example is further explained in the subsequent sections):
```

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md [hunk 16]
```diff
@@ -210,3 +201,177 @@ public link(parentContext: Web3Context) {
...
+		customRpcMethods: CustomRpcMethodsPlugin;
+	}
+}
+
+export { Web3Context };
+```
+
+The following represent what your plugin users would see:
+
+![web3 context augmentation side effect](./assets/web3_context_augmentation_side_effect.png 'web3Context augmentation side effect')
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_authors.md]: Here also, I would like to know what the others think about this also :smile:. Especially our other approver: @avkos 
```suggestion
![web3 context augmentation](./assets/web3_context_augmentation.png 'web3Context augmentation')
```

(The file name would need to be also modified accordingly.)

**Review by Muhammad-Altabba — COMMENTED** [review]
Looks great!
However, there is tiny stuff I found in the second round.
Thanks,

**Comment by Muhammad-Altabba** [comment]
> > > And regarding the file named `web3_context_augmentation_side_effect.png`. What I see in the screenshot is the result of the augmentation **main** effect, which we are seeking. And using the words `side_effect` would probably make some confusion. As we call a "side effect", usually, on something that is not intended to happen, but happened due to some side effect. So, I suggest to simply call the file `web3_context_augmentation.png`.
> > 
> > 
> > @Muhammad-Altabba The screenshot is in the context of IntelliSense still showing the plugin's functionality even if the plugin hasn't been registered to an instance of the Web3 class - This is not intended functionality, but is a repercussion of the plugin augmenting `Web3Context`
> 
> Ok, I got your point now smile But I think about it this way: The screenshot shows the direct-effect, that is the main-effect, and that is almost-only-effect of the Module augmentation. And showing that the IntelliSense will happen, regardless of calling `registerPlugin`, is something not to be confused by what is the main-effect and what is the side-effect of the Module augmentation. So, actually having the IntelliSense work is the intended effect of the module augmentation. The intended effect means we do the Modul augmentation for only one reason and it works for that reason. We do the module augmentation to have IntelliSense and it works as intended. And having the IntelliSense work regardless of calling `registerPlugin` is something that does not alter what is the Module augmentation's intended purpose. So, I think you meant to call it `web3_context_augmentation_regardless_of_calling_registerPlugin.png` but that would be a long name though smile and `web3_context_augmentation.png` is good enough, I think. I think that we can keep the file name as you named it. I just think that it would confuse the developers who read its name later. Or what do you think?

Sorry I missed that it is not only the file name. So, I added suggestions here: https://github.com/web3/web3.js/pull/5621#discussion_r1030835220 and here https://github.com/web3/web3.js/pull/5621#discussion_r1031118627 . And also asked for the opinion of our other approver @avkos :smile: .

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md (hunk 14 above)
Comment (Spacesai1or) [reply docs/docs/guides/web3_plugin_guide/plugin_authors.md]: There's no need, it's a copy paste error and they should all be CustomRpcMethodsPlugin

**Comment by avkos** [comment]
> > > > And regarding the file named `web3_context_augmentation_side_effect.png`. What I see in the screenshot is the result of the augmentation **main** effect, which we are seeking. And using the words `side_effect` would probably make some confusion. As we call a "side effect", usually, on something that is not intended to happen, but happened due to some side effect. So, I suggest to simply call the file `web3_context_augmentation.png`.
> > > 
> > > 
> > > @Muhammad-Altabba The screenshot is in the context of IntelliSense still showing the plugin's functionality even if the plugin hasn't been registered to an instance of the Web3 class - This is not intended functionality, but is a repercussion of the plugin augmenting `Web3Context`
> > 
> > 
> > Ok, I got your point now smile But I think about it this way: The screenshot shows the direct-effect, that is the main-effect, and that is almost-only-effect of the Module augmentation. And showing that the IntelliSense will happen, regardless of calling `registerPlugin`, is something not to be confused by what is the main-effect and what is the side-effect of the Module augmentation. So, actually having the IntelliSense work is the intended effect of the module augmentation. The intended effect means we do the Modul augmentation for only one reason and it works for that reason. We do the module augmentation to have IntelliSense and it works as intended. And having the IntelliSense work regardless of calling `registerPlugin` is something that does not alter what is the Module augmentation's intended purpose. So, I think you meant to call it `web3_context_augmentation_regardless_of_calling_registerPlugin.png` but that would be a long name though smile and `web3_context_augmentation.png` is good enough, I think. I think that we can keep the file name as you named it. I just think that it would confuse the developers who read its name later. Or what do you think?
> 
> Sorry I missed that it is not only the file name. So, I added suggestions here: [#5621 (comment)](https://github.com/web3/web3.js/pull/5621#discussion_r1030835220) and here [#5621 (comment)](https://github.com/web3/web3.js/pull/5621#discussion_r1031118627) . And also asked for the opinion of our other approver @avkos 😄 .

You are right. we do not need to create confusing names for images and also in documentation. This is an effect that we expect. So it shouldn't be a "side-effect"

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md (hunk 15 above)
Comment (Spacesai1or) [reply docs/docs/guides/web3_plugin_guide/plugin_authors.md]: Addressed by [this](https://github.com/web3/web3.js/pull/5621/commits/921ab8816ea0578e8b51abd23e1f2a22df24fbef) commit

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md (hunk 16 above)
Comment (Spacesai1or) [reply docs/docs/guides/web3_plugin_guide/plugin_authors.md]: Addressed by [this](https://github.com/web3/web3.js/pull/5621/commits/921ab8816ea0578e8b51abd23e1f2a22df24fbef) commit

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md (hunk 14 above)
Comment (Spacesai1or) [reply docs/docs/guides/web3_plugin_guide/plugin_authors.md]: Addressed by [this](https://github.com/web3/web3.js/pull/5621/commits/63ee5f86c0a47d3eeac40c4dc101caf4a1540b46) commit

File: docs/docs/guides/web3_plugin_guide/plugin_authors.md [hunk 17]
```diff
@@ -210,3 +201,177 @@ public link(parentContext: Web3Context) {
...
+		customRpcMethods: CustomRpcMethodsPlugin;
+	}
+}
+
+export { Web3Context };
+```
+
+The following represent what your plugin users would see:
+
+![web3 context augmentation side effect](./assets/web3_context_augmentation.png 'web3Context augmentation side effect')
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_plugin_guide/plugin_authors.md]: Here is a tiny missed modification, I think:
```suggestion
![web3 context augmentation](./assets/web3_context_augmentation.png 'web3Context augmentation')
```

**Comment by Muhammad-Altabba** [comment]
There is something not related to this MR. But I think it could be done here as this PR is about documentation refactoring. 

I think that we would better use in our samples `Web3EthPluginBase` instead of `Web3PluginBase`. Because I expect that most of the users will simply use `Web3EthPluginBase` (in which they can still add custom RPC methods in addition to the standard Ethereum ones. But it makes the standard ones just easily available). If you agree, I think it is Ok to include the replacement in this MR. And if not, I would create an issue to discuss this.
Or what do you think?

**Review by Muhammad-Altabba — APPROVED** [review]
Thanks @spacesailor24 for your great effort.
I am approving this MR. However, I hope you address this https://github.com/web3/web3.js/pull/5621#issuecomment-1328723169 if you find it fits within :wink:

**Comment by Spacesai1or** [comment]
> There is something not related to this MR. But I think it could be done here as this PR is about documentation refactoring.
> 
> I think that we would better use in our samples `Web3EthPluginBase` instead of `Web3PluginBase`. Because I expect that most of the users will simply use `Web3EthPluginBase` (in which they can still add custom RPC methods in addition to the standard Ethereum ones. But it makes the standard ones just easily available). If you agree, I think it is Ok to include the replacement in this MR. And if not, I would create an issue to discuss this. Or what do you think?

@Muhammad-Altabba Extending `Web3EthPluginBase` means that the plugin is going to send the standard ETH JSON RPC calls to a provider using the provided `requestManager`. In most cases, users probably aren't going to do this and will instead import the JSON RPC wrapper they desire from either `web3-eth` or `web3-rpc-methods` packages. However, I did include a new [Extending Web3EthPluginBase](https://github.com/web3/web3.js/pull/5621/commits/dadd6c21a42c325ff06c4e40ff1bccea8db6dcc2) section
